### PR TITLE
Copies references to data structures to avoid extra ARC traffic.

### PR DESCRIPTION
**Pull request by saeta**
Note: this change (by itself) does not significantly reduce ARC traffic, but in concert
with `UnmanagedBuffer` (#41), we see a ~15x performance improvement in
the parallelFor benchmark.

Performance for `NonBlockingThreadPool: parallel for, one level`:
```
master (b5ab4b6)                                   336813.5 ns  ±  65.93 %   6526        
managed buffers (9e31153)                         5062117.0 ns  ±  16.01 %    280         
this change (1ae770d)                              210097.0 ns  ±  58.77 %   9347        
tmp branch with managed buffers + this change       22078.0 ns  ± 942.16 %   8510        
```

Note: the variability in timing increases substantially as performance improves. This is (I believe) due to the lower frequency of stealing.

### CODE REVIEW

**Comment by saeta** [comment]
https://github.com/saeta/penguin/issues/33

**Comment by dabrahams** [comment]
> #33

A couple of words describing the relationship you're establishing by mentioning that issue here would be appreciated.

File: Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift [hunk 1]
```diff
@@ -462,35 +466,40 @@ fileprivate final class PerThreadState<Environment: ConcurrencyPlatform> {
...
+      if selectedThreadId >= totalThreadCount {
+        selectedThreadId -= totalThreadCount
       }
     }
     return nil
   }
 
   func spin() -> Task? {
-    let spinCount = pool.threads.count > 0 ? Constants.spinCount / pool.threads.count : 0
+    let spinCount = totalThreadCount > 0 ? Constants.spinCount / totalThreadCount : 0
```
Comment (dabrahams) [inline Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift:502]: This change looks suspect.  Elsewhere you have replaced `pool.totalThreadCount` with `totalThreadCount` but here you're replacing `pool.threads.count` which is presumably a different quantity (if it isn't, I question the value of `pool.totalThreadCount`!)

Comment (saeta) [reply Sources/PenguinParallel/NonblockingThreadPool/NonBlockingThreadPool.swift:502]: Good eye! This should be the number of worker threads and not the `totalThreadCount`. Thanks!

**Comment by dabrahams** [comment]
@saeta You didn't address my suggestion to turn on tsan.  Building parallel stuff and not testing with tsan seems really super inadvisable.  At least file an issue?

**Pull request merged**